### PR TITLE
Add more tree generators

### PR DESCRIPTION
**Pull request by jimmikaelkael**
_Edited by turt2live_

**Related Links:**
- Ticket: #76 

---

It adds the following trees:
- Acacia
- Dark Oak
- Swamp tree
- Jungle bush
- Big jungle tree
- Tall redwood tree (those who have just a few leaves)
- Mega redwood tree
#### Test Plugin:
- jar: https://drive.google.com/file/d/0B0gnjB7PN1h3d2YxM25MWVZ6ZmM/view?usp=sharing
- source available here: https://gist.github.com/jimmikaelkael/3d3448223c64870bbecd

Acacia

Dark Oak

Swamp tree

Jungle tree

Tall redwood

Mega redwood tree

### CODE REVIEW

**Comment by dequis** [comment]
~~Compare link for the interesting part of this PR: https://github.com/jimmikaelkael/Glowstone/compare/b766e17...more-trees~~

**Comment by jimmikaelkael** [comment]
Bot failing ? I just added a commit but it still WIP...

**Comment by jimmikaelkael** [comment]
@turt2live ready to get this PR reviewed ^^

File: src/main/java/net/glowstone/generator/objects/trees/AcaciaTree.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+    public AcaciaTree(Random random, Location location, BlockStateDelegate delegate) {
+        super(random, location, delegate);
+        setHeight(random.nextInt(3) + random.nextInt(3) + 5);
+    }
+
+    @Override
+    public boolean canPlaceOn() {
+        final BlockState state = delegate.getBlockState(loc.getBlock().getRelative(BlockFace.DOWN).getLocation());
+        if (state.getType() != Material.GRASS
```
Comment (turt2live) [inline src/main/java/net/glowstone/generator/objects/trees/AcaciaTree.java]: Any particular reason why the use of 2-line if statements is used throughout? In many cases you can shorten it to a single return...

File: src/main/java/net/glowstone/generator/objects/trees/AcaciaTree.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean generate() {
+
+        // check height range
+        if (!canHeightFit()) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/generator/objects/trees/AcaciaTree.java]: Combining the next 3 if statements would be suitable (same for the rest of the PR)

File: src/main/java/net/glowstone/generator/objects/trees/AcaciaTree.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            }
+        }
+
+        // block below trunk is always dirt
+        delegate.setTypeAndRawData(loc.getWorld(), loc.getBlockX(), loc.getBlockY() - 1, loc.getBlockZ(), Material.DIRT, 0);
+
+        return true;
+    }
+
+    private void setLeaveAt(int x, int y, int z) {
```
Comment (turt2live) [inline src/main/java/net/glowstone/generator/objects/trees/AcaciaTree.java]: This seems badly named

File: src/main/java/net/glowstone/generator/objects/trees/CocoaTree.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                delegate.setTypeAndData(loc.getWorld(), loc.getBlockX(), loc.getBlockY() + y, loc.getBlockZ() - 1, Material.VINE, new Vine(BlockFace.SOUTH));
+            }
+            if (random.nextInt(3) != 0
+                    && delegate.getBlockState(loc.getWorld(), loc.getBlockX(), loc.getBlockY() + y, loc.getBlockZ() + 1).getType() == Material.AIR) {
+                delegate.setTypeAndData(loc.getWorld(), loc.getBlockX(), loc.getBlockY() + y, loc.getBlockZ() + 1, Material.VINE, new Vine(BlockFace.NORTH));
+            }
+        }
+    }
+
+    protected void addVinesOnLeaves() {
```
Comment (jimmikaelkael) [inline src/main/java/net/glowstone/generator/objects/trees/CocoaTree.java:37]: @jimmikaelkael Put this above the private methods please.

Comment (bendem) [inline src/main/java/net/glowstone/generator/objects/trees/CocoaTree.java:37]: that is awkward

Comment (jimmikaelkael) [inline src/main/java/net/glowstone/generator/objects/trees/CocoaTree.java:37]: I was at work at that time so I left this comment for me to remember to change it later. Thank you to be rude...

Comment (turt2live) [inline src/main/java/net/glowstone/generator/objects/trees/CocoaTree.java:37]: Yea @bendem, don't be rude. Your comment is not required.

Comment (bendem) [inline src/main/java/net/glowstone/generator/objects/trees/CocoaTree.java:37]: Not meant to be rude, sorry :crying_cat_face:

**Comment by SpaceManiac** [comment]
Merged in bac4337a7c8f6230f25975d36cd2b1a51a6709e1.

**Pull request closed without merging**